Default to local leaderboard if selected one is unavailable (has no scores)

wooster0 opened the pull request:
This is supposed to resolve https://github.com/ppy/osu/discussions/23713

I tested this locally and by submitting scores globally.
If I have a global leaderboard available with scores on it and start a beatmap, it stays global and I can compete against the global leaderboard.
If I select say the country leaderboard but I don't have the osu!supporter tag, it defaults to local.
If global is unavailable, it also defaults to local. And so on.

One thing that this PR probably doesn't yet capture is https://github.com/ppy/osu/discussions/23713#discussioncomment-6098781 because I don't quite understand what @peppy meant by that.
Perhaps the idea was to use global as an even better fallback to local if available? Is that @peppy meant?
Or was the idea that if by chance the global leaderboard gains a score while the beatmap is played, it switches to global?

Commit:
Default to local leaderboard if selected one is unavailable

bdach commented:
> One thing that this PR probably doesn't yet capture is [#23713 (comment)](https://github.com/ppy/osu/discussions/23713#discussioncomment-6098781) because I don't quite understand what @peppy meant by that.
> Perhaps the idea was to use global as an even better fallback to local if available? Is that @peppy meant?

I *think* this may have meant that the fallback should be to global leaderboard if the selected beatmap has one, and to local otherwise...? Shouldn't be too hard to do, probably could use beatmap online status for this (something like `status >= BeatmapOnlineStatus.Ranked`). But probably better to confirm that first because I'm not 100% sure.

bdach commented on osu.Game/Online/Leaderboards/Leaderboard.cs:
```diff
@@ -360,6 +360,14 @@ private void setState(LeaderboardState state)
             }
         }
 
+        /// <summary>
+        /// Whether the leaderboard is displaying scores (at least one).
+        /// </summary>
+        public bool HasScores()
+        {
+            return state == LeaderboardState.Success;
+        }
```
I don't think there's much point to this method. I'd just expose `state` as publicly readable (but _not_ settable) at this point.

bdach commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -116,6 +120,12 @@ protected override bool OnStart()
 
+            // Default to local leaderboard if the currently selected leaderboard doesn't have scores or is unavailable
+            // perhaps because it requires sign in, requires the beatmap to be ranked, etc.
+            bool isLeaderboardSelected = config.Get<PlayBeatmapDetailArea.TabType>(OsuSetting.BeatmapDetailTab) != PlayBeatmapDetailArea.TabType.Details;
+            if (isLeaderboardSelected && !playBeatmapDetailArea.Leaderboard.HasScores())
+                config.SetValue(OsuSetting.BeatmapDetailTab, PlayBeatmapDetailArea.TabType.Local);
```
Rather than `.Get<>()` and then `.SetValue()`, I'd just get `config.GetBindable<>()`, store that to a field, and then read from/write to the setting via the bindable.

wooster0 commented:
Applied the suggestions and tested that everything still works.

> But probably better to confirm that first because I'm not 100% sure.

Now waiting for a response from @peppy.

Commit:
Apply code review suggestions

Commit:
Also default to local if details tab is selected

Originally I made it so that if the details tab was selected in song
select, there's no in-game leaderboard visible but you can already hide
the in-game leaderboard with the tab key.

peppy commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -116,6 +123,12 @@ protected override bool OnStart()
+            // perhaps because it requires sign in, requires the beatmap to be ranked, etc.
+            // Also default to local if the details tab is selected so the leaderboard with the new score is visible after returning to song select.
+            if (playBeatmapDetailArea.Leaderboard.State != LeaderboardState.Success || beatmapDetailTab.Value == PlayBeatmapDetailArea.TabType.Details)
+                beatmapDetailTab.Value = PlayBeatmapDetailArea.TabType.Local;
```
This is going to overwrite the user setting, ie. change their tab permanently. Is that the intention? I think this could be a bit weird if a user is going down their list of songs and one doesn't have global leaderboards, therefore triggers the tab to change (forever).

wooster0 replied:
Yes, that is the intention. Doesn't this only trigger when a beatmap is started instead of the list just being scrolled? I tested a bit locally with friend leaderboard selected (none of my beatmaps have this available) and just by scrolling through my list the selected tab didn't change.

peppy replied:
You will find most players use down or right arrow keys to listen and view previews of each map in their carousel.

wooster0 replied:
I didn't know. I usually just scroll with my tablet and don't really touch my keyboard unless I'm typing.

I still don't quite understand what this has to do with this though because whichever way you navigate through your beatmaps in song select, unless you don't actually start a map this code is never going to be executed because this is in `OnStart`.
I tested it locally and pressing the arrow keys to navigate through the beatmaps never changes the leaderboard to the local one if the currently selected one is unavailable.

peppy replied:
Hmm, doing it on start is pretty weird. I'd argue that users will be quite confused by this. Will have to think about it a bit.

peppy commented:
This was closed due to a fork deletion, but it's safe to say that the underlying systems have changed to the point that the solution here would require a rewrite anyway.

This may be something we want to look into still as a future effort. The case where a user is at song select, has global leaderboards selected but is viewing a pending beatmap as one example. Showing local scores would likely be preferred to showing nothing.
